applications/CompressiblePotentialFlowApplication/custom_processes/apply_far_field_process.cpp

```diff
@@ -0,0 +1,129 @@
+    std::size_t num_threads = omp_get_max_threads();
+    std::vector<double> min_projections(num_threads, std::numeric_limits<double>::epsilon());
+    std::vector<std::size_t> nodes_id_list(num_threads, 0);
+
+    //Find minimum node in the direction of the free stream velocity in openmp.
+    double distance_projection=0.0;
+    #pragma omp parallel for firstprivate (distance_projection)
+    for (int i = 0; i < static_cast<int>(mrModelPart.Nodes().size()); i++) {
+        auto it_node = mrModelPart.NodesBegin() + i;
+        std::size_t thread_id = omp_get_thread_num();
+        const auto& r_coordinates = it_node->Coordinates();
+        distance_projection = inner_prod(r_coordinates, mFreeStreamVelocity);
+        if (distance_projection < min_projections[thread_id]){
+            min_projections[thread_id] = distance_projection;
+            nodes_id_list[thread_id] = it_node->Id();
+        }
+    }
+
+    // Find minimum across all threads
+    std::size_t minimum_node_thread_id = 0;
+    for (std::size_t i_thread = 0; i_thread<num_threads; i_thread++){
+        if (min_projections[i_thread] < min_projections[minimum_node_thread_id]){
+            minimum_node_thread_id = i_thread;
+        }
+    }
+
+    // Storing final result in member variable
+    mpReferenceNode = mrModelPart.pGetNode(nodes_id_list[minimum_node_thread_id]);
+}
+
+void ApplyFarFieldProcess::AssignFarFieldBoundaryConditions()
+{
+    double velocity_projection = 0.0;
+    array_1d<double,3> normal;
+    array_1d<double,3> aux_coordinates;
+
+    #pragma omp parallel for firstprivate(velocity_projection, normal, aux_coordinates)
+    for (int i = 0; i < static_cast<int>(mrModelPart.Conditions().size()); i++) {
+        auto it_cond = mrModelPart.ConditionsBegin() + i;
+        auto& r_geometry = it_cond->GetGeometry();
+        // Computing normal
+        r_geometry.PointLocalCoordinates(aux_coordinates, r_geometry.Center());
+        normal = it_cond->GetGeometry().Normal(aux_coordinates);
+
+        velocity_projection = inner_prod(normal, mFreeStreamVelocity);
+        if (velocity_projection < 0.0) {
+            AssignDirichletFarFieldBoundaryCondition(r_geometry);
+        }
+        else {
+            AssignNeumannFarFieldBoundaryCondition(*it_cond);
+        }
+
+    }
+}
+
+void ApplyFarFieldProcess::AssignDirichletFarFieldBoundaryCondition(Geometry<NodeType>& rGeometry)
+{
+    // Fixing nodes in the domain that are part of the inlet, and initializing its value
+    // according to the free stream velocity.
+    for (std::size_t i_node = 0; i_node < rGeometry.size(); i_node++){
+        array_1d<double,3> relative_coordinates = rGeometry[i_node].Coordinates() - mpReferenceNode->Coordinates();
+        double inlet_potential = inner_prod(relative_coordinates, mFreeStreamVelocity);
+        rGeometry[i_node].Fix(VELOCITY_POTENTIAL);
+        rGeometry[i_node].FastGetSolutionStepValue(VELOCITY_POTENTIAL) = inlet_potential + mReferencePotential;
```

Comment:
I am not sure this is threadsafe since two different conditions from different threads may try to write into the same node at the same time. You might need to lock the node first. (Not sure whether this is also necessary for the Fix)

Reply:
You are right, I didnt think about it!